### PR TITLE
WIP: MSC3277: Scheduled messages

### PR DESCRIPTION
**Pull request by ara4n**

### CODE REVIEW

File: proposals/3277-scheduled-messages.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+time in order to minimise embargo leaks.
+
+The server must set the `origin_server_ts` of the message is set to be its
+`at` time, given in practice this is the label used for clients to display
+the intended delivery timestamp of a message.  It must also set `scheduled:
+true` on the event (not the contents) in order to let servers know to treat
+it differently for history visibility purposes, and so the sender's clients
+all know that it's scheduled.  The recipients can then also see the message
+is scheduled. (A future MSC may define how to suppress this field from
+recipients if desired).
```
[the remark this answers is not in the document]
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:44]: pretty sure it doesn't need a new room version. we're just special-casing what events are allowed to go down /sync and /messages on the CS API, rather than changing the federation semantics?

[the remark this answers is not in the document]
Comment (turt2live) [reply proposals/3277-scheduled-messages.md:44]: Having thought about this for a whole 5 minutes, I think we can just go with the approach alluded to in the numbered list, but that list is founded on false pretenses anyways. 

The concern was that if a server joined the room such that the history visibility would cause the remote server to receive a redacted event, however this isn't possible: even if the history is restricted as `invited` or `joined`, the remote server will still receive the unredacted event over `/send` when it's due for delivery, and can then impose local history visibility checks on `/sync` and friends for its users. The origin server could probably be smart enough to serve an unredacted copy of the event if it knew the caller would have received it, but that's largely out of scope for this MSC. The only case a server *should* be backfilling events it received over `/send` is when the database got trimmed for some reason. Another MSC can address the statefulness of Matrix.

All told: auth rules are *not* changed here, as they can still be enforced appropriately. Servers are simply modifying their history visibility checks, which are checked at time of exposure (to clients and other servers) and not time of sending, as the auth rules check at time of sending/receipt.

File: proposals/3277-scheduled-messages.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+to send the message at the right time, which feels fragile (or encourages a
+longlived serverside client/bot to be run, which could put E2EE at risk).
+
+## Security considerations
+
+This MSC allows for spammers to potentially pre-plan their attacks for
+off-peak hours when the moderators may not be present. It also potentially
+allows a user to stack up a large number of pending events which all get sent
+throughout the day as a disruption attempt. Servers should impose harsh
+limits to the number of events a user may schedule, such as a maximum of 3
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:160]: > such as a maximum of 3 pending events

per room?

File: proposals/3277-scheduled-messages.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+This is a simplified proposal derived from Travis' [original draft](https://docs.google.com/document/d/1vFbQuCnCNURBRs4_zukxiKLW_p2Ka7Ogo42oahpXnz4),
+which aims to incorporate all the feedback gathered there.
+
+## Proposal
+
+A new query parameter, `at`, is added to the [/send](https://matrix.org/docs/spec/client_server/r0.6.1#put-matrix-client-r0-rooms-roomid-send-eventtype-txnid)
+and [PUT /state](https://matrix.org/docs/spec/client_server/r0.6.1#put-matrix-client-r0-rooms-roomid-state-eventtype-statekey)
+endpoints to indicate when the origin server should aim to send the event
+(milliseconds since the unix epoch). The message is added to the room DAG
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:20]: How does this play with the prev_events of the events that follow directly after? Or will this event be set aside to have no event reference it in prev_events?

Removing it from future prev_events prevents (hah) many issues wrt PL/State/Auth validation after sending, because if you effectively try to re-validate it at the unveiling time, and the event soft-fails, you're creating a hole in the DAG tree (deliberately)

File: proposals/3277-scheduled-messages.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+than servers and we want clients to be as simple as possible.
+
+Separately, there's a trade-off between whether you send the queued message
+out over federation at the time it's created or at the scheduled time.  If
+you send at create-time, it means that the origin server doesn't need to
+still be running at the scheduled time. On the other hand, it means that you
+are much more likely to break embargos, as one-person homeservers will get
+the message ahead of schedule.  We've gone for the more privacy preserving
+option (send over federation at the scheduled time).  Plus it acts as an
+incentive for people to keep the origin server running ;P
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:139]: > Plus it acts as an incentive for people to keep the origin server running

This may be impossible in a P2P scenario, where a user's device might not allow the app to wake up at the scheduled time, or the device may not be online entirely. This may be solved by a P2P equivalent though, where the device sends the event to whatever trusted forwarding server, which then releases the event at the scheduled time.

File: proposals/3277-scheduled-messages.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+room keys for past messages) to avoid the messages being undecryptable by
+future devices.
+
+The megolm session must be a one-off used just for this scheduled message
+(meaning that as it's shared for future participants in the room, it only exposes
+the session used to encrypt that particular message rather than any surrounding
+non-scheduled ones).
+
+E2EE implementations must not discard 'unused' megolm sessions, given they
+may be shared long in advance of the scheduled message.
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:73]: This may be used to DoS or spam other devices with bogus megolm sessions, as now the equivalent of "just trust me" is happening between these devices. Some malicious actor could start sending these sessions to other devices in a room, and then never release a scheduled event, and i assume that sessions are somewhat expensive/non-trivial to reason with/store on client devices, so a malicious actor could fill up someone's megolm database with bogus sessions for which no events have been decrypted with.

File: proposals/3277-scheduled-messages.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+keyshare request and let any device in the room share to the newcomer if they
+believe the newcomer should be entitled to decrypt the message. This
+introduces the attack that a malicious server admin could add a fake
+user/device to their local copy of a room, and could use this to exfiltrate
+keys from bystanders - but this may be an acceptable tradeoff to improve
+the chances of the message being decryptable in the distant future.
+Alternatively, MLS may solve this(??)
+
+Users who send a scheduled message and are then kicked out of a room will have
+no way of being able to cancel their scheduled message.  This could either be
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:105]: Wouldn't the scheduled message fail if the user isn't in the room? As technically they wouldn't have "access" to the room per PL rules normally, so they certainly cant inject a scheduled message in there?

File: proposals/3277-scheduled-messages.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+is scheduled. (A future MSC may define how to suppress this field from
+recipients if desired).
+
+For rooms with non-shared history visibility, servers must special-case the
+visibility of scheduled events based on the membership of the room at the
+scheduled time, rather than based on its DAG position (which reflects the
+creation time).
+
+The sender's clients will want to track the unsent scheduled messages in its
+various rooms, such that the sender can cancel them by redacting them, or
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:52]: Note: wording this "redacting" initially confused me, until i realised that it's about an event which has already been inserted in the DAG. Still though, wouldn't it be somewhat of an information leak if suddenly 5 events would appear in a room, 2 redacted, and another 2 signalling edits of the one message that isn't redacted?

I'd say that leaving the original injected event (in conjunction with "not having future prev_events refer to it" as said before) to be mutable until sent would be better.

File: proposals/3277-scheduled-messages.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+limits to the number of events a user may schedule, such as a maximum of 3
+pending events, to reduce the potential for abuse while still maintaining an
+amount of usability for common scenarios.
+
+Users could also try to schedule many events at once or schedule such that
+they all get sent at once - servers should apply rate limiting on both the
+scheduling and sending sides to limit the user’s ability to spam a room.
+
+This proposal relies on absolute timestamps, and so for it to work sensibly
+servers need to have an accurate (e.g. NTP-synced) clock.
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:169]: I could comment that this is unattainable, and that it somewhat breaks a promise of matrix, in that it doesn't necessarily rely on earth-ran clocks, but that it runs on the "this happens after that" principle

That said, scheduled messages trust the clock anyways, so that point is a bit moot, what i instead then argue for is to introduce a mechanic which synchronises or compares the *client* and server clocks, in that if a client clock is ahead or behind, the server's clock can offset that and deliver the event on a time that the client still expects it. 

This may need some further fiddling, but then only two clocks are trusted, and they're not trusted to be synchronised, only to be monotonic and same in speed. The target time will then only (at most) deviate by how long it took for the client to send the event to the server.

**Review by ShadowJonathan — COMMENTED** [review]
Some notes, but this is already fairly solid :+1: 

Imo it also would need a JSON example of a schedule message, but that's minor.

File: proposals/3277-scheduled-messages.md (hunk 3 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:20]: the event would sit effectively soft-failed until the time comes to actually send it, so it would not be referred to by future prev_events. will clarify.

Comment (ara4n) [reply proposals/3277-scheduled-messages.md:20]: (have completely reworded this now)

File: proposals/3277-scheduled-messages.md (hunk 4 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:138]: good point. for P2P we probably want to change the behaviour so that the client sends it (anonymised) to a long-lived relay rather than relying on the origin. will clarify.

File: proposals/3277-scheduled-messages.md (hunk 6 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:104]: excellent point :D

File: proposals/3277-scheduled-messages.md (hunk 7 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:51]: you can't make the original message mutable, as that would change its event ID. instead you have to use normal redaction semantics. a smart server could conceivably avoid sending out the redacted drafts of the msg by setting the replacement draft to be a sibling of the redaction rather than a child in the DAG, but i don't think we need to spec that here.  I'll put it as a footnote.

File: proposals/3277-scheduled-messages.md (hunk 8 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:164]: the only bit of this spec which requires synchronised clocks is the "display the scheduled message at the right absolute time", which is an inevitability.  So i don't think we need this clarification.

File: proposals/3277-scheduled-messages.md (hunk 5 above)
Comment (ara4n) [reply proposals/3277-scheduled-messages.md:73]: this is true irrespective of this proposal, but yeah, will note as a security consideration.

**Comment by ara4n** [comment]
@ShadowJonathan thanks for excellent review - have incorporated all feedback (i think).

File: proposals/3277-scheduled-messages.md [hunk 9]
```diff
@@ -50,10 +49,11 @@ creation time).
 
 The sender's clients will want to track the unsent scheduled messages in its
 various rooms, such that the sender can cancel them by redacting them, or
-edit them (optionally redacting earlier versions, providing private
-drafting).  We propose using the same mechanism as for tracking 'starred'
-(aka 'favourited' or 'flagged') messages, to avoid sprouting two different
-APIs for almost identical functionality.  This is deferred to a future MSC.
+edit them (optionally redacting earlier versions <sup id="a1">[1]
+(#f1)</sup>, providing private drafting).  We propose using the same
+mechanism as for tracking 'starred'(aka 'favourited' or 'flagged') messages,
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:54]:
```suggestion
mechanism as for tracking 'starred' (aka 'favourited' or 'flagged') messages,
```
(wraparound whitespace typo)

File: proposals/3277-scheduled-messages.md [hunk 10]
```diff
@@ -101,18 +102,46 @@ keys from bystanders - but this may be an acceptable tradeoff to improve
...
+{
+  "content": {
+    "body": "a note to my future self",
+    "msgtype": "m.text"
+  },
+  "origin_server_ts": 1636197454551,
+  "sender": "@matthew:matrix.org",
+  "type": "m.room.message",
+  "unsigned": {
+    "age": 391
```
Comment (ShadowJonathan) [inline proposals/3277-scheduled-messages.md:135]: Would it make sense to add `"scheduled": true` to unsigned here?

**Review by ShadowJonathan — APPROVED** [review]
LGTM as-is

File: proposals/3277-scheduled-messages.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+creation time in order to minimise embargo leaks.  If the event can't be
+re-authed (e.g. due to the sender no longer being in the room, or no longer
+has sufficient PL to send the event) then the event is soft-failed.  As such,
+the event effectively appears as if it had been received over federation
+after a federation delay.  This means all event types can be scheduled in
+this manner, including state events.  Normal event ID and redaction behaviour
+applies.  
+
+The server must set the `origin_server_ts` of the message is set to be its
+`at` time, given in practice this is the label used for clients to display
```
Comment (kevincox) [inline proposals/3277-scheduled-messages.md:37]: This should only be done if the `at` timestamp is "reasonable". For example if the timestamp is in the past the server should probably still set `origin_server_ts` as it would if the event was not scheduled. For example this could occur if a scheduled message was sent while offline and the device did not come online until after the send time.

File: proposals/3277-scheduled-messages.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+(#f2)</sup>
+
+Another alternative would be to use an API shape where you put a field on the
+event contents (e.g. `m.pts` for presentation timestamp) to tell servers and
+clients alike that a given message is scheduled.  However, this ends up with
+two timestamps on the event (confusing), complicates E2EE further
+(don't encrypt the `m.pts`), and makes it harder to hide from recipients
+whether a message was scheduled.  So we've stuck with `?at=`.
+
+We could avoid all the e2ee and history visibility problems if the sending
```
Comment (kevincox) [inline proposals/3277-scheduled-messages.md:178]:
```suggestion
We could avoid all the E2EE and history visibility problems if the sending
```

Nit: Consistency.

File: proposals/3277-scheduled-messages.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+user simply queued the message in their clients until it was ready to send.
+However, this of course means they have to keep an always on client somewhere
+to send the message at the right time, which feels fragile (or encourages a
+longlived serverside client/bot to be run, which could put E2EE at risk).
+
+## Security considerations
```
Comment (kevincox) [inline proposals/3277-scheduled-messages.md:184]: Once the client queues the E2EE event they have no way of stopping the server from sending it. Of course there is always a race condition here but the server could publish at any time even if the user changes their mind. This could be somewhat mitigated by clients throwing away that key ASAP which could prevent anyone from decrypting the message.

For example consider that Alice wants to share information to Mallory, but only in 2 weeks time. If Mallory could break into Alice's server Mallory can force-send that message beforehand (or just exfiltrate that message).

However I think that in general this is the required tradeoff to allow sending the message while the client is not online.

**Review by kevincox — APPROVED** [review]

File: proposals/3277-scheduled-messages.md (hunk 11 above)
Comment (ShadowJonathan) [reply proposals/3277-scheduled-messages.md:37]: The "server should clamp past time stamps" can get finicky if the clocks of the server and the client aren't synchronized.

Comment (kevincox) [reply proposals/3277-scheduled-messages.md:37]: I agree but I don't think the server always trusts the client enough to pick arbitrary timestamps. This could lead to confusing behaviour. I get that this is possible over federation with a malicious server anyways but it seems like this confusing behaviour should be avoided when it is simple to do a pretty good job. We are already trusting the server clock for when to publish the message so it seems that the server slot is expected to be vaguely in sync anyways.

TL;DR if the field is called `origin_server_ts` the server should have the final say. I don't think it is actually too important that we require precise semantics but I think we should allow the server a clean way to avoid sending known-invalid timestamps. If you don't like clamping we can consider rejecting the message upon being queued. However this is really pushing the timestamp-guessing onto the client which is more flexible but probably suboptimal for most use-cases.

**Comment by JMoVS** [comment]
I was almost going to write up my own proposal but then discovered this. thumbs up - this would be a really really great feature to have in business contexts!

File: proposals/3277-scheduled-messages.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+received.  Therefore we add a `m.forward_until: timestamp` field to the
+`m.room_key` to-device message to indicate the session's lifetime. The
+timestamp should be that of the scheduled event; users who join the room
+after that timestamp should not be able to read the message.
+
+E2EE implementations must not discard 'unused' megolm sessions for scheduled
+messages (i.e. megolm sessions with `m.forward_until` fields), given they may
+be shared long in advance of the scheduled message.
+
+## Possible issues
```
Comment (turt2live) [inline proposals/3277-scheduled-messages.md:90]: This MSC allows the server to set `origin_server_ts` to a future date, but if it doesn't then there's a bit of a privacy leak: https://twitter.com/soren_iverson/status/1785308725362016505 (not a real feature)

**Comment by mpeter50** [comment]
> I was almost going to write up my own proposal but then discovered this. thumbs up - this would be a really really great feature to have in business contexts!

Not only in a business context, I think. I often use such messages to set up reminders using an other messaging platform.

I know this is not the proper way to do reminders, but the todo software "collection" I use across my devices often only syncs at half an hour intervals or such (and since the operation is heavy enough, I dont want it to be any more requent), so for short term reminders this is more reliable.
But then this also solves reminders for an arbitrary group of friends who are available over Matrix (and any bridged platforms), which is an additional feature even if that todo stack was quick to sync.
